### ISSUE TITLE
Overlapping Leave - Query Issue

### ISSUE THREAD

**Issue by marct83** [timeoff-management/application#249]
Having some trouble with the Overlapping Leave query. If I try and put a leave in for a day before a day that already has a leave. I get error:

"An error occured when user 73 try to create a leave request: Error: Overlapping booking!"

However if I try and put in a leave for the day after a day that already has a leave there is no issue.

I have figured out the issue. The Query it runs is the following:

SELECT "id", "status", "employee_comment", "approver_comment", "decided_at", "date_start", "day_part_start", "date_end", "day_part_end", "createdAt", "updatedAt", "userId", "approverId", "leaveTypeId" FROM "Leaves" AS "Leave" WHERE ("Leave"."userId" = 73 AND (("Leave"."status" != 3 AND "Leave"."status" != 5) AND ("Leave"."date_start" BETWEEN '2018-03-07' AND '2018-03-08' OR "Leave"."date_end" BETWEEN '2018-03-07' AND '2018-03-08')));

If I'm putting a leave in for 2018-03-07 and there is a leave on 2018-03-08. It will return the leave on 2018-03-08. The between is inclusive that is why it returns 2018-03-08.

If I'm putting a leave in for 2018-03-09 and there is a leave on 2018-03-08. It will not return anything as the query would be the following and there is no leave on 2018-03-10:

SELECT "id", "status", "employee_comment", "approver_comment", "decided_at", "date_start", "day_part_start", "date_end", "day_part_end", "createdAt", "updatedAt", "userId", "approverId", "leaveTypeId" FROM "Leaves" AS "Leave" WHERE ("Leave"."userId" = 73 AND (("Leave"."status" != 3 AND "Leave"."status" != 5) AND ("Leave"."date_start" BETWEEN '2018-03-09' AND '2018-03-10' OR "Leave"."date_end" BETWEEN '2018-03-09' AND '2018-03-10')));

We are using a Postgres database. I'm unsure if the Between function is different between database types and the dates are treated differently in mysql?

This Between could possibly be changed to:

("Leave"."date_start" >= '2018-03-07' 
AND "Leave"."date_end" < '2018-03-08' 
OR "Leave"."date_end" >= '2018-03-07' 
AND "Leave"."date_end" < '2018-03-08')

**Comment by vpp** [comment]
@marct83 ensure that you have correct time zone configured in the company settings.

**Comment by vpp** [comment]
and make sure your server is running with UTC time on

**Comment by vpp** [comment]
@marct83 really strange..

I've checked that scenario on cloud version account and all is good there.

Are you sure that when you are adding new leave start date and end dates are the same?

**Comment by marct83** [comment]
Yes start and end are the same date. Unsure if Postgres has anything to do with this.

**Comment by vpp** [comment]
dates in BETWEEN '2018-03-07' AND '2018-03-08' are start and end date submitted from front end (as far as i remember)

**Comment by vpp** [comment]
ignore previous comment:

here is how the condition is done:

`133   this.validate_overlapping = function(new_leave_attributes) {
134     var this_user = this;
135
136     var days_filter = {
137       $between : [
138         new_leave_attributes.from_date,
139         moment.utc(new_leave_attributes.to_date)
140           .add(1,'days').format('YYYY-MM-DD'),
141       ],
142     };

**Comment by marct83** [comment]
Ya I just tested and that is what it is doing.

**Comment by vpp** [comment]
@marct83 I am going to have a look at this thing later as it is really weird.

Although I cannot reproduce the problem over here, I do not see how it works here.

(it probably would take awhile as I am going to be busy next week)

**Comment by marct83** [comment]
I will test it in a dev mysqllite db with every setup the same and report back.

**Comment by vpp** [comment]
@marct83 one more thing: which time zone you are in?

**Comment by vpp** [comment]
and one more thing which date format you are using?

**Comment by vpp** [comment]
I also assume that you use the most recent version

**Comment by marct83** [comment]
We are in the America/Regina Timezone. We pointed it to Mysqllite database and it works fine. We changed it back to Postgres and its back to not working. Also tried it on a different machine/server and same issues. Version is the version before the theme colors went in. So fairly recent.

What do you mean by date format?

**Comment by vpp** [comment]
Thanks for sharing.

About date format: I meant the field in company settings, like DD/MM/YY or MM/DD/YY etc

**Comment by marct83** [comment]
We tried this:

   var days_filter = {
      $between : [
        new_leave_attributes.from_date,
        moment.utc(new_leave_attributes.to_date)
          .endOf('day').format('YYYY-MM-DD'),
      ],
    };

It seems to have fixed it. Also tested it with multiple scenarios and it seems good.

**Comment by shellhouse** [comment]
@marct83 Your code update above really saved the day in our case, thank you.  Running on MySQL and experienced the same issue.

@vpp Would it be possible to test and add to the main project?  Everything was the same in our case, with GMT/UTC being correct.  If it doesn't break other configurations it'd be very helpful for future updates.  Thanks so much!
